### PR TITLE
VTK I/O Support

### PR DESCRIPTION
**Pull request by joshichaitanya3**
This PR adds a `vtk` module to Morpho (along with tests and documentation). 

This module provides support for I/O of Fields and / or Meshes using the VTK File Format. The data is stored in the Legacy VTK format. 

The documentation can be accessed by running 

`help vtk`

The module currently only supports I/O of scalar fields or vector fields that live on the vertices ( fields of shape `[1,0,0]`). Support for tensorial fields and fields on cells can be added.

It also adds a `parser` module, which, for now, only contains a `Tokenizer` class that is used by the `VTKImporter`.

### CODE REVIEW

File: morpho5/modules/vtk.morpho [hunk 1]
```diff
@@ -0,0 +1,495 @@
...
+
+        self.maxg = self.mesh.maxgrade()
+
+    }
+
+    // Write the preamble for the file. 
+    _preamble(file) {
+
+        file.write("# vtk DataFile Version 3.0 ") 
+        file.write("Exported using Morpho https://github.com/Morpho-lang/morpho ") // This line can be anything the user wants.
```
Comment (ConduitDan) [inline morpho5/modules/vtk.morpho:107]: // This line can be anything the user wants.

does the user have a way of specifying this?
or is the user here us?

File: test/vtk/ensurevtkfilename.morpho [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    
+    vtkE.export(filename)
+
+    vtkI = VTKImporter(filename)
+
+    var m2 = vtkI.mesh()
+
+    print m2 
+}
+// expect: <Mesh: 2 vertices>
```
Comment (ConduitDan) [inline test/vtk/ensurevtkfilename.morpho:27]: You're not actually checking the desired behavior here.
If you compare the vtkE's file name you should read in from something like
`expectedfilenames = [ "data.vtk", "data.case2.vtk", "data.case3.0.vtk"]`

or use File to check for existence

you might also consider adding a test file name ending in .vtk, I think that case is missed

File: test/vtk/export_and_import_mesh.morpho [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import vtk
```
Comment (ConduitDan) [inline test/vtk/export_and_import_mesh.morpho:1]: This test looks redundant with ensurevtkfilename.

Do you want to be checking m1 against m2?

File: test/vtk/export_and_import_no_fieldname.morpho [hunk 4]
```diff
@@ -0,0 +1,40 @@
+// Testing whether exporting and importing a field with no field name
+// works. In this case, the fieldname will be either "scalars" or "vectors"
```
Comment (ConduitDan) [inline test/vtk/export_and_import_no_fieldname.morpho:2]: Is this behavior captured in the doc?

**Review by ConduitDan — COMMENTED** [review]
Looks good!

What was the reasoning behind using the Legacy format rather than (I assume) the contemporary one?

Testing question: did you try importing any of your .vtk files into VTK? to ensure the formatting is correct? I don't know if that something we want in our main testing suit but we could have some rarely running tests that ensure our compatibility with VTK

File: morpho5/modules/parser.morpho [hunk 5]
```diff
@@ -0,0 +1,44 @@
+
```
[the remark this answers is not in the document]
Comment (softmattertheory) [reply morpho5/modules/parser.morpho:1]: The parser here is a placeholder for a parser module. Once we have that, we can get rid of this!

**Comment by softmattertheory** [comment]
@ConduitDan  are we ready to merge this, or do we need fixes?

**Comment by ConduitDan** [comment]
I think as long as @joshichaitanya3 has indeed checked that the files created work with VTK we can go ahead and merge.

I think the test suite for this needs a small fix or two but that can be taken up afterwards (we should just create a test issue for it before merging)

**Comment by joshichaitanya3** [comment]
Thanks for the review @ConduitDan and sorry I haven't gotten to this yet.

I will add a test with an externally validated VTK file this afternoon. I'll try to address the tests issues, but I'm okay with merging and creating a separate issue for tracking.

**Comment by ConduitDan** [comment]
Sounds good,

As long as you have checked manually that your output is acceptable to VTK for a verity of scenarios I'm fine with the merge (with the automated tests to be added later)

**Pull request merged**